dipy/reconst/fwdti.py

```diff
@@ -0,0 +1,713 @@
+            warnings.simplefilter("ignore")
+            w = 1/(se + C**2)
+            # The weights are normalized to the mean weight (see p. 1089):
+            w = w/np.mean(w)
+
+    # Return the weighted residuals:
+    with warnings.catch_warnings():
+        warnings.simplefilter("ignore")
+        return np.sqrt(w * se)
+
+
+def nlls_fit_tensor(design_matrix, data, fw_params=None, Diso=3e-3,
+                    weighting=None, sigma=None, cholesky=False,
+                    f_transform=True):
+    """
+    Fit the water elimination tensor model using the non-linear least-squares.
+
+    Parameters
+    ----------
+    design_matrix : array (g, 7)
+        Design matrix holding the covariants used to solve for the regression
+        coefficients.
+
+    data : array ([X, Y, Z, ...], g)
+        Data or response variables holding the data. Note that the last
+        dimension should contain the data. It makes no copies of data.
+
+    fw_params: ([X, Y, Z, ...], 14), optional
+           A first model parameters guess (3 eigenvalues, 3 coordinates
+           of 3 eigenvalues, the volume fraction of the free water
+           compartment, and the estimate of the non diffusion-weighted signal
+           S0). If the initial fw_paramters are not given, function will use
+           the WLS free water elimination algorithm to estimate the parameters
+           first guess.
+           Default: None
+
+    Diso : float, optional
+        Value of the free water isotropic diffusion. Default is set to 3e-3
+        $mm^{2}.s^{-1}$. Please ajust this value if you are assuming different
+        units of diffusion.
+
+    weighting: str, optional
+           the weighting scheme to use in considering the
+           squared-error. Default behavior is to use uniform weighting. Other
+           options: 'sigma' 'gmm'
+
+    sigma: float, optional
+        If the 'sigma' weighting scheme is used, a value of sigma needs to be
+        provided here. According to [Chang2005]_, a good value to use is
+        1.5267 * std(background_noise), where background_noise is estimated
+        from some part of the image known to contain no signal (only noise).
+
+    cholesky : bool, optional 
+        If true it uses cholesky decomposition to insure that diffusion tensor
+        is positive define.
+        Default: False
+
+    f_transform : bool, optional
+        If true, the water volume fractions is converted during the convergence 
+        procedure to ft = arcsin(2*f - 1) + pi/2, insuring f estimates between
+        0 and 1.
+        Default: True
+
+    Returns
+    -------
+    nlls_params: for each voxel the eigen-values and eigen-vectors of the
+        tissue tensor and the volume fraction of the free water compartment.
+    """
+    # Flatten for the iteration over voxels:
+    flat_data = data.reshape((-1, data.shape[-1]))
+
+    # Use the WLS method parameters as the starting point if fw_params is None:
+    if fw_params==None:
```

Comment:
I'm making some suggestions here. Mostly related to inputting an initial guess from a WLS fit if you run it separately.
```
if np.any(fw_params) is None:
    fw_params = wls_fit_tensor(design_matrix, flat_data,  Diso=Diso)
else: #ETP
    fw_params = fw_params.reshape((-1, 14)) 
```